Promote ServiceLBNodePortControl feature to Beta

hanlins opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Promotes ServiceLBNodePortControl feature gate to Beta. Integration API tests were added #96610

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
This PR is based on https://github.com/kubernetes/kubernetes/pull/100019/, completed the tests related missing pieces.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Promote the feature ServiceLBNodePortControl to Beta
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

andrewsykim committed:
promote ServiceLBNodePortControl to beta

Signed-off-by: Andrew Sy Kim <kim.andrewsy@gmail.com>

hanlins committed:
Update the related tests

1. add AllocateLoadBalancerNodePorts fields in specs for validation test cases
2. update fuzzer
3. in resource quota e2e, allocate node port for loadbalancer type service and
   exceed the node port quota

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/test pull-kubernetes-e2e-kind-ipv6

hanlins commented:
/assign @uablrek @thockin
cc @andrewsykim @maplain

hanlins commented:
/test pull-kubernetes-unit

hanlins commented:
All tests passed, PTAL @thockin.
/assign @liggitt as apimachinery e2e test is touched.

hanlins commented:
/test pull-kubernetes-integration

hanlins commented:
/test pull-kubernetes-e2e-kind

aojea commented:
/sig network
I'm afraid that this missed the 1.21 code freeze deadline.
I don't know if you are still on time to request an exception :thinking: 
https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md
https://www.kubernetes.dev/resources/release/

fedebongio commented:
/remove-sig api-machinery

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10540,13 +10546,15 @@ func TestValidateServiceCreate(t *testing.T) {
 			name: "valid type - loadbalancer",
 			tweakSvc: func(s *core.Service) {
 				s.Spec.Type = core.ServiceTypeLoadBalancer
+				s.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(true)
```
We probably want a case where this is false

hanlins replied:
Will add a negative case, thanks for the remainder.

hanlins replied:
Added case where `AllocateLoadBalancerNodePorts` is set to `false` and another negative case that leaves `AllocateLoadBalancerNodePorts` unset.

aojea replied:
don't we have to enable the featuregate during the test?

```
	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceLBNodePortControl)()
```

hanlins replied:
I think in this PR the feature is enabled by default already (see [here](https://github.com/kubernetes/kubernetes/blob/b1b8f049e7aa757d48559ffc2690271203b0cbf4/pkg/features/kube_features.go#L849)).

aojea replied:
hmm, I always thought that the test shouldn't depend on the global state, let me ask @liggitt , he always know all these things better than anybody.
Is just for learning purposes :smiley:

andrewsykim replied:
This is validation though, so the feature gate shouldn't impact the logic?

aojea replied:
https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/pkg/apis/core/validation/validation.go#L4375-L4377

hanlins replied:
My understanding is all features are loaded from `defaultKubernetesFeatureGates` to the global `DefaultMutableFeatureGate` in `init`https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/pkg/features/kube_features.go#L740-L742
And `ServiceLBNodePortControl` is now enabled in `defaultKubernetesFeatureGates`.
https://github.com/kubernetes/kubernetes/blob/b1b8f049e7aa757d48559ffc2690271203b0cbf4/pkg/features/kube_features.go#L849
Also `DefaultFeatureGate` is the same as `DefaultMutableFeatureGate`
https://github.com/kubernetes/kubernetes/blob/204ff6caeb8229e6d12b54809108e7f32c19502f/staging/src/k8s.io/apiserver/pkg/util/feature/feature_gate.go#L30-L32

So later when the validation happens, it's using the global `DefaultFeatureGate`, so it should know that `ServiceLBNodePortControl` is being enabled.

aojea replied:
yeah, not a big deal, but my question was more related to know if we do want to test with and without feature gate ... but I think that I understand know Andrew's point, it will not be possible since we clear in strategy IIUIC
disregard my comment, sorry

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -13049,6 +13081,7 @@ func TestValidateServiceUpdate(t *testing.T) {
 			tweakSvc: func(oldSvc, newSvc *core.Service) {
 				oldSvc.Spec.Type = core.ServiceTypeClusterIP
 				newSvc.Spec.Type = core.ServiceTypeLoadBalancer
+				newSvc.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(true)
```
How about a case where we change this ?  I thought we only wanted to allow it to change when type becomes LB?  But I can't seem to find that

hanlins replied:
Do you mean leaving `newSvc.Spec.AllocateLoadBalancerNodePorts` as empty here?

hanlins replied:
Oh I see what you mean, added tests that changes `AllocateLoadBalancerNodePorts` from `true` to `false` and the other way around.

hanlins committed:
Add tests for LB type service

1. create LB type svc with nodeport allocation set to false
1. create LB type svc with nodeport allocation unset
3. update LB type svc's nodeport allocation field

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

caesarxuchao commented:
/remove-sig api-machinery

hanlins committed:
Fix the beta release version.

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

hanlins commented:
/retest

hanlins commented:
/test pull-kubernetes-integration

fedebongio commented:
/remove-sig api-machinery

fedebongio commented:
/triage accepted

andrewsykim commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10540,13 +10546,30 @@ func TestValidateServiceCreate(t *testing.T) {
+			numErrs: 0,
+		},
+		{
+			name: "valid type - loadbalancer",
```
`valid type - loadbalancer with allocateLoadBalancerNodePorts=false"`

andrewsykim commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10540,13 +10546,30 @@ func TestValidateServiceCreate(t *testing.T) {
 			name: "valid type - loadbalancer",
```
`valid type - loadbalancer with allocateLoadBalancerNodePorts=true"`

andrewsykim commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -91,6 +90,7 @@ var _ = SIGDescribe("ResourceQuota", func() {
 		ginkgo.By("Creating a ResourceQuota")
 		quotaName := "test-quota"
 		resourceQuota := newTestResourceQuota(quotaName)
+		resourceQuota.Spec.Hard[v1.ResourceServices] = resource.MustParse("2")
```
What's the reason for this change?

hanlins replied:
Check this [failure case](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100412/pull-kubernetes-e2e-kind-ipv6/1373025220083847168) in PR history. This test created two services and expects the third service creation to be failed due to the quota issue. However it didn't fail so I set the quota to 2, then it worked. It was 10 originally by default.

aojea replied:
@pacoxu you fixed the bug related to this https://github.com/kubernetes/kubernetes/commit/15867d9e8a1faf007f6df563c26a9b5e8744b2a1#diff-e1a46008c4ddbb9f83858cb22ce56ca0f820b78142a1e8abbbb11aa59127a09b , can you PTAL

thockin replied:
do we have resolution on this?  I do not understand this change

hanlins replied:
Hi @thockin, you can see that as we initialize the quota using `newTestResourceQuota`, in line 1561, by default we can create 10 services:
https://github.com/kubernetes/kubernetes/blob/49a60be8ab5578097dc2dd2b851213d9f8ca1105/test/e2e/apimachinery/resource_quota.go#L1557-L1578
And read through the test, it basically expect the first 2 service (one of type cluster IP and the other of type node port) to be created successfully, and expect the third creation (which happens to be a service of LB type) to fail. 

What I remember when I did the tests months back was, before this PR, the third service(the LB type one) creation failed, but it's because the `ServiceLBNodePortControl` feature was not turned on by default. It has nothing to do with quota here. And since this PR enabled that feature by default, the third service was successfully created which failed the test. So it's a bug in the test case and all I should do is just set the quota to the correct value which is 2.

andrewsykim commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -111,7 +111,7 @@ var _ = SIGDescribe("ResourceQuota", func() {
 
 		ginkgo.By("Not allowing a LoadBalancer Service with NodePort to be created that exceeds remaining quota")
-		loadbalancer := newTestServiceForQuota("test-service-lb", v1.ServiceTypeLoadBalancer, false)
+		loadbalancer := newTestServiceForQuota("test-service-lb", v1.ServiceTypeLoadBalancer, true)
```
What's the reason for this change?

hanlins replied:
This seems not relevant, whether we create NP or not for this service, it should fail. Will change it back.

hanlins replied:
Revisited the case and sorted it out. Actually this is a bug in the test. The creation request was rejected because the `allocateLoadBalancerNodePorts` was turned off by default (thus each LB type service will allocate a node port and exceed quota and failed). Now as the feature is enabled by default, the creation will success if we don't allocate a node port to the service. We need to set it to true here so that the node port quota will exceed. Btw I've added more tests for this `allocateLoadBalancerNodePorts` feature in a separate conformance test so it's not interfering with the conformance tests written since 1.16.

thockin replied:
We should have a case where we pass true and it fails and another where we pass false and it does not fail.

hanlins commented:
/test pull-kubernetes-e2e-kind-ipv6

hanlins commented:
/test pull-kubernetes-e2e-kind

hanlins commented:
/test pull-kubernetes-e2e-kind

hanlins commented on pkg/apis/core/fuzzer/fuzzer.go:
```diff
@@ -518,6 +521,11 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 			case core.ServiceAffinityNone:
 				ss.SessionAffinityConfig = nil
 			}
+			if feature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
```
That seems reasonable, will fix it, thanks!

andrewsykim replied:
(I don't actually know the answer, asking for clarification)

hanlins commented:
/test pull-kubernetes-e2e-kind

uablrek commented:
This lgtm. Assuming the validation assures that the pointer is never nil.

aojea commented:
@hanlins @andrewsykim is this ready for final review?
 I see it was updated recently, and extensively reviewed before
It will be good if we start to get in these "smaller" ones now, before the big features at the end of the release

hanlins commented:
> @hanlins @andrewsykim is this ready for final review?
> I see it was updated recently, and extensively reviewed before
> It will be good if we start to get in these "smaller" ones now, before the big features at the end of the release

Yes, it's ready, let me know if any further change is needed!

aojea commented:
/priority important-soon
ping @thockin

thockin commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10517,6 +10517,7 @@ func TestValidateServiceCreate(t *testing.T) {
 			name: "valid load balancer protocol UDP 1",
 			tweakSvc: func(s *core.Service) {
 				s.Spec.Type = core.ServiceTypeLoadBalancer
+				s.Spec.AllocateLoadBalancerNodePorts = utilpointer.BoolPtr(true)
```
Rather than doing this in every tweak, it should go in `makeValidService()`.  The tweaks are emphasizing what is DIFFERENT in each test case.

hanlins replied:
The `AllocateLoadBalancerNodePorts` should only be set if the service type is `LoadBalancer`:
https://github.com/kubernetes/kubernetes/blob/7e23c5367dd6dc5b1eb304b395efebb1e01485a8/pkg/apis/core/types.go#L3733-L3737

And in the `makeValidService`, it's creating a service of type `ClusterIP`:
https://github.com/kubernetes/kubernetes/blob/4d38d21880a6f9c988fe5a1408ed71725f8084d8/pkg/apis/core/validation/validation_test.go#L9972-L9984

So for each `LoadBalancer` case, we have to do the change in the tweak function I think.

thockin commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -1737,6 +1737,12 @@ func newTestReplicaSetForQuota(name, image string, replicas int32) *appsv1.Repli
 
 // newTestServiceForQuota returns a simple service
 func newTestServiceForQuota(name string, serviceType v1.ServiceType, allocateLoadBalancerNodePorts bool) *v1.Service {
```
All three places that call this pass `false` for `allocateLoadBalancerNodePorts`.

Can you go thru that and make sure it tests quota for type=NodePort *and* NodePorts allocated in type=LoadBalancer when this is true *and not* NodePorts in type=LoadBalancer when this is false?

hanlins replied:
Not sure if this is what you're expecting, added the happy path for LB type service creation both with and w/o node port, and retain the negative case for resource quota exceed test.

hanlins replied:
Ahh, had a few tries and I just realized that, the `allocateLoadBalancerNodePorts` feature is still Beta and is not enabled in the GA feature only conformance tests (namely `pull-kubernetes-conformance-kind-ga-only-parallel`), so it was failing consistently. That said, in the current resource quota conformance tests, I think we need to assume that the `allocateLoadBalancerNodePorts` feature is not enabled and for each LB type service created, a node port will be allocated. I reverted the new conformance test, and we can enable the `allocateLoadBalancerNodePorts` specific test once the feature is GAed.

thockin replied:
It worries me that we have more work to do when we GA - it seems likely to get forgotten.  Can we add the code now, but explicitly check the gate for `PreRelease = featuregate.GA`.  That way it starts testing as soon as we mark it GA and when we remove the gate, we will also have to remove that check.

hanlins replied:
Hi @thockin, I think `defaultKubernetesFeatureGates` is not exposed
https://github.com/kubernetes/kubernetes/blob/38f012320fa138260dca09e135ed386026a061bc/pkg/features/kube_features.go#L726
so it's hard to get maturity of a specific feature. By checking feature gate https://github.com/kubernetes/kubernetes/blob/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L86-L96
we can check if a feature is enabled, but we can't get its maturity. So for now the test will be skipped if only GAed features are enabled and will be covered if all beta features are enabled. We have both testbed in our pipeline so both scenarios are covered.

BenTheElder replied:
conformance tests may not skip under any circumstances.
non-GA features need to be tested in not-a-conformance-test.

hanlins replied:
Followed suggestions from @bentheelder and @liggitt, for now keep the test as an e2e test. Later once we promote the feature to GA, we can promote the test to a conformance test.

BenTheElder replied:
even if the feature were GA: we usually start by adding a test if missing and then wait for evidence that the test is stable + reliable on some existing CI before promoting it to a test everyone must pass when certifying their clusters 😅

hanlins replied:
Followed suggestion from @aojea, moved the test from e2e to an integration test.

thockin commented:
Let's remember to update https://github.com/orgs/kubernetes/projects/10 when this merges

hanlins commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

hanlins commented:
/test pull-kubernetes-verify

hanlins commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

liggitt commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -136,6 +137,95 @@ var _ = SIGDescribe("ResourceQuota", func() {
+		Create another ClusterIP type Service. Its creation MUST fail as the service resource quota is exceeded.
+		Delete the Services. Deletion MUST succeed and resource usage count against the Service object MUST be released from ResourceQuotaStatus of the ResourceQuota.
+	*/
+	framework.ConformanceIt("should create a ResourceQuota and capture the life of a load balancer type service with allocateLoadBalancerNodePorts.", func() {
```
start this as a regular e2e test... once the feature graduates to GA and this demonstrates stability, it can be promoted to a conformance test

hanlins replied:
I see, just converted it to a e2e test, thanks!

hanlins commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

hanlins commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

hanlins commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

andrewsykim commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -136,6 +137,109 @@ var _ = SIGDescribe("ResourceQuota", func() {
 
+	// TODO: convert the test to conformance test once the loadbalancer node port control feature is GAed
+	ginkgo.It("should create a ResourceQuota and capture the life of a load balancer type service with allocateLoadBalancerNodePorts.", func() {
+		if !feature.DefaultFeatureGate.Enabled(features.ServiceLBNodePortControl) {
```
IIRC, I don't think this will work because test binaries do not register feature gates

hanlins replied:
Thanks @andrewsykim! Actually the feature gate is not the main problem, the main issue I hit was, the LB type service which allocates node port can't be deleted (this behavior is consistent). I've put more details on [slack thread](https://kubernetes.slack.com/archives/C09QYUH5W/p1625009880371900), for now I doubt if it's a bug on service deletion. I tried create and delete the LB type service on my own environment, and it was working fine:
```
➜  ~ kubectl apply -f svc-lb.yaml
deployment.apps/nginx-deployment unchanged
service/nginx-lb created
➜  ~ kubectl get svc nginx-lb
NAME       TYPE           CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE
nginx-lb   LoadBalancer   10.100.64.51   <pending>     8080:31466/TCP   8s
➜  ~ kubectl get svc nginx-lb -o wide
NAME       TYPE           CLUSTER-IP     EXTERNAL-IP   PORT(S)          AGE   SELECTOR
nginx-lb   LoadBalancer   10.100.64.51   <pending>     8080:31466/TCP   11s   app=nginx
➜  ~ time kubectl delete svc nginx-lb
service "nginx-lb" deleted
kubectl delete svc nginx-lb  0.09s user 0.04s system 73% cpu 0.184 total
➜  ~ kubectl get svc nginx-lb
Error from server (NotFound): services "nginx-lb" not found
```
Not sure what happended in the test infra that the service cannot be cleaned up. @aojea do you have any idea on how to debug the `pull-kubernetes-e2e-gce-ubuntu-containerd` environment? From the [debug log](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100412/pull-kubernetes-e2e-gce-ubuntu-containerd/1409994251747463168#1:build-log.txt%3A37141) I can see that the LB type service is not removed, which is different from my local test results. Anything special with that environment?

aojea commented on test/e2e/apimachinery/resource_quota.go:
```diff
@@ -135,6 +135,79 @@ var _ = SIGDescribe("ResourceQuota", func() {
 	})
 
+	// TODO: convert the test to conformance test once the loadbalancer node port control feature is GAed
+	ginkgo.It("should create a ResourceQuota and capture the life of a load balancer type service with allocateLoadBalancerNodePorts.", func() {
```
This can be done with an integration test since we are not exercising the data path, it will be also much cleaner since we can make it trable driven.

andrewsykim replied:
^ agree, if we can put this test into https://github.com/kubernetes/kubernetes/blob/master/test/integration/service/loadbalancer_test.go it would be ideal. But there's a question about whether this is a Service test or a ResourceQuota test

hanlins replied:
Yeah I would say the intention was to test the node port quota but it requires to provision a LB type service. I can try add it in the quota integration test, it seems there's no service related tests for quota yet.

hanlins replied:
Moved the e2e test to quota integration test, now it's passing.

hanlins committed:
Add integration test for LB node port control

Signed-off-by: Hanlin Shi <shihanlin9@gmail.com>

thockin commented on pkg/registry/core/service/storage/rest_test.go:
```diff
@@ -2150,9 +2156,10 @@ func TestServiceRegistryExternalTrafficHealthCheckNodePortUserAllocation(t *test
+			Selector:                      map[string]string{"bar": "baz"},
+			SessionAffinity:               api.ServiceAffinityNone,
+			Type:                          api.ServiceTypeLoadBalancer,
+			AllocateLoadBalancerNodePorts: utilpointer.BoolPtr(true),
```
Sadly a change just went it which will force you to rebase this (this file in particular).  You can add AllocateLoadBalancerNodePorts to the pkg/api/service/testing/make.go `MakeService` and get all of these for free.

thockin reviewed (commented):
I'm not super familiar with the integration framework anymore.  It seems like a lot of setup, but the test cases are robust.

/approve

hanlins commented:
/test pull-kubernetes-unit

hanlins commented:
/test pull-kubernetes-e2e-gce-alpha-features

hanlins commented:
/test pull-kubernetes-unit

aojea commented:
/lgtm

pull-kubernetes-e2e-gce-alpha-features 

seems unrelated, it is failing in other prs

BenTheElder commented:
alpha features is broken, https://github.com/kubernetes/kubernetes/issues/103440

andrewsykim commented:
/retitle Promote ServiceLBNodePortControl feature to Beta

thockin reviewed (commented):
I have a related PR - I will fix my nits.
